src/Chapter3.hs

```diff
@@ -1006,6 +1129,118 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data KAction = KAttack | KPotion | KSpell
+
+data K = MkK
+  { kHealth :: Health
+  , kAttack :: Attack
+  , kDefense :: Defense
+  , kActions :: [KAction]
+  , usedSpell :: Bool
+  , usedPotion :: Bool
+  }
+
+mkK :: Health -> Attack -> Defense -> [KAction] -> K
+mkK h a d k =
+  MkK { kHealth = h
+      , kAttack = a
+      , kDefense = d
+      , kActions = cycle k
+      , usedSpell = False
+      , usedPotion = False
+      }
+
+data MAction = MAttack | MFlee
+data M = MkM
+  { mHealth :: Health
+  , mAttack :: Attack
+  , mActions :: [MAction]
+  }
+
+mkM :: Health -> Attack -> [MAction] -> M
+mkM h a m =
+  MkM { mHealth = h
+      , mAttack = a
+      , mActions = cycle m
+      }
+
+data Turn = TAttack Damage | TIdle | TFlee
+
+class Fighter a where
+  damage :: a -> Damage
+  getDamaged :: a -> Damage -> a
+  isDead :: a -> Bool
+  nextAction :: a -> (Turn, a)
+
+instance Fighter K where
+  damage :: K -> Damage
+  damage k = Damage (a + (if usedSpell k then 1 else 0))
+    where Attack a = kAttack k
+
+  getDamaged :: K -> Damage -> K
+  getDamaged k (Damage d) =
+    let Health kHealth' = kHealth k in
+    let Defense kDefense' = kDefense k in
+    k { kHealth = Health (min (kHealth' + kDefense' + (if usedPotion k then 1 else 0) - d) kHealth') }
+
+  isDead :: K -> Bool
+  isDead k = h <= 0
+    where Health h = kHealth k
+
+  nextAction :: K -> (Turn, K)
+  nextAction k =
+    let action = head $ kActions k
```

Comment:
`head` is a partial function and it fails on empty lists. Even though you use `cycle` to create values of your data types, nobody guarantees that people will always use `cycle` to create list of actions.

You can restrict the fact that the list is not empty by using the `NonEmpty` data type:

* http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-List-NonEmpty.html#t:NonEmpty

If you want, you can even model your types to make this case impossible. This pattern is called "Making illegal states unrepresentable". You can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable

For example, if you really want to work with only infinite lists, you can create a separate data type for this:

```haskell
data Stream a = Stream a (Stream a)
```

and a helper function:

```haskell
cycleToStream :: NonEmpty a -> Stream a
```

Laziness in Haskell allows creating infinite data structures :exploding_head: